### DISCUSSION TITLE
Static generation with a custom server

### DISCUSSION THREAD

**Discussion by marnixhoh** [vercel/next.js#23121] [Help] [answered]
Hi,

I am currently working on migrating a CRA app to Next.js and I need some help with understanding the following:

My application has a custom Node (express) backend. When I was using CRA, I would just serve the build statically using the same Express server, which is serving the API. So my first thought is to use this same Express server to serve the Next.js app as well. I.e. using a custom server as explained here: `https://nextjs.org/docs/advanced-features/custom-server`.

Now, that resource is saying the following: `A custom server will remove important performance optimizations, like serverless functions and Automatic Static Optimization.`. It seems like not being able to use serverless functions is not a problem in my case, since I am using a custom backend anyways. But I am not sure if the same is true for Automatic Static Optimization.

Does this mean that I can still use static generation by explicitly providing getStaticProps? I.e. it just won't be automatically inferred anymore if static prerendering is possible? Because if this is the case, it doesn't appear to be that big of a problem for me, as I can easily explicitly provide getStaticProps.

And finally, does it actually make sense in my case to use a custom server to serve the Next.js app or is this some sort of anti-pattern?

Would love to hear your thoughts! Any help or suggestions is greatly appreciated! Thanks :D

**Comment by akellbl4** [comment, marked as answer]
Hi! If you would like to stick with SSG it's better to start your next.js app as an independent process and just use API from the Express app. Think about Next.js as frontend and Express as backend and keep them separately.
In my opinion, you have no reason to wrap your Next.js in a custom server.

**Reply by leerob** [reply]
Yep, I would agree with this! Good answer 😄

**Reply by marnixhoh** [reply]
Ah ok. Thank you so much for your help @akellbl4 & @leerob!

Just for my understanding: even though it might not be necessary in my case to use a custom server, if I would use one anyway, are there any downsides? As I said in my post, it seems to me like there won't be any. But perhaps I'm missing something?

**Reply by akellbl4** [reply]
As you wrote in the first message, you will lose `Automatic Static Optimization`.
I'm not sure if it's true. I've tried to create an application with a custom server. I put two pages in that app. The first is without data fetching at all and the second one is with `getStaticProps`. They both were generated and served as static HTML pages. The only difference I can see is that I have my own node server on top of it. It could cause some difference from using Next.js without a custom server due to my server logic and that's all.

@leerob am I right? After my experiments, I'm not sure if docs are correct or what means under `ASO`.

**Reply by marnixhoh** [reply]
@akellbl4 Yes that is exactly what I thought as well!

Plus even if `ASO` wouldn't work using a custom server, one could still just explicitly use `getStaticProps`, alleviating the need for `ASO` right?

**Reply by akellbl4** [reply]
Yes. `ASO` is helpful because you can don't think about static optimization, and Next.js will take care of it. When you care about it and use `getStaticProps` explicitly on your pages they will be statically optimized even if you use a custom server.

As I wrote before, in my opinion, it's better to keep backend and frontend separate from each other because you can:
- control resources that are taken by Next.js;
- get errors separately from your backend;
- debug your app easily;

Also,  you can have a fully working frontend in the case when your backend is down.

**Reply by marnixhoh** [reply]
Gotcha! That makes sense. Thank you so much for your help and your suggestion to keep the front- and backend separate.

I currently use AWS so I will have to research what the best way is to host such a setup. If you happen to have experience with this, I'd love your opinion #23146.

Again, thank you so much :)

**Reply by kristijorgji** [reply]
There are a lot of reasons to use a custom server like:
1. There is no way to properly monitor a nextjs server side application now, log exceptions and have good visibility. Tools like Newrelic that I tried understand that ok respose is 500 but don't capture any stack trace or anything. With a custom server you can use a global exception handler. Example: you have a getServerSideProps method that makes some api call to agregate data. During the call something goes south and exception is thrown, in a proper designed application you are notified right away with full stack trace of error. Currently I see no such option for nextjs server side without ejecting a custom server and registering global exception handler -> send to sentry/newrelic the stack trace
2. With a custom server you can have endpoints that return xml as well, example /sitemap.xml generate on fly or from cached in disk anywhere etc. Same with json responses and other responses for mobile association files and other places. Nextjs default server allows us to only return html rendered via the react components under `/pages/somepagename.tsx`
3. full power, partial rendering in html if you want shared redis cache across next cluester etc etc

So question here is and I am interested as well: How to preserve static generation functionality while using a custom server ? My assumption is that works out of box and will test that now

**Reply by kristijorgji** [reply]
Update:
I did a naive test with custom server.ts (to setup typescript server.ts followed https://github.com/vercel/next.js/tree/canary/examples/custom-server-typescript)
and I think static generation worked fine.

Test was like this:
I built then started the compiled dist/server.js
I opened one test page /posts/222 that is using in the pages/posts/[id].tsx getStaticProps and getStaticPaths method with revalidate: 7200

I checked api requests, first time api was called to get the post. Then on the next calls it was not called, and I assume next used the generated html+json props until 7200 seconds will expire.

I also saw at folder 
`.next/server/pages/posts/` two files created `222.html` and `222.json`

So to my humble assumption which I need confirmation from next team seem to work fine and am not losing SSG capability with custom server.
